File: sdk/typescript/src/sdk.ts
/**
 * TigerScan TypeScript SDK
 * 
 * Comprehensive TypeScript SDK for TigerScan API and TigerSmartChain blockchain
 */

import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
import { ethers, BigNumber, Contract, Wallet } from 'ethers';

// Types
export enum ChainId {
  TIGER_MAINNET = 1,
  TIGER_TESTNET = 5,
  ETHEREUM_MAINNET = 1,
  ETHEREUM_SEPOLIA = 11155111,
}

export interface Block {
  number: number;
  hash: string;
  parentHash: string;
  nonce: string;
  sha3Uncles: string;
  logsBloom: string;
  transactionsRoot: string;
  stateRoot: string;
  receiptsRoot: string;
  miner: string;
  difficulty: number;
  totalDifficulty: number;
  size: number;
  gasLimit: number;
  gasUsed: number;
  timestamp: number;
  transactionCount: number;
  transactions: string[];
  uncles: string[];
}

export interface Transaction {
  hash: string;
  nonce: number;
  blockHash: string;
  blockNumber: number;
  transactionIndex: number;
  from: string;
  to: string;
  value: string;
  gasPrice: string;
  gasLimit: number;
  gasUsed: number;
  input: string;
  status: number;
  timestamp: number;
}

export interface TransactionReceipt {
  transactionHash: string;
  blockHash: string;
  blockNumber: number;
  cumulativeGasUsed: number;
  gasUsed: number;
  contractAddress: string | null;
  status: number;
  logs: Log[];
}

export interface Log {
  address: string;
  topics: string[];
  data: string;
  blockNumber: number;
  logIndex: number;
  transactionIndex: number;
}

export interface Token {
  address: string;
  name: string;
  symbol: string;
  decimals: number;
  totalSupply: string;
  price: string;
  marketCap: string;
  volume24h: string;
  holders: number;
  transfers: number;
}

export interface TokenHolder {
  address: string;
  balance: string;
  percent: number;
}

export interface NFT {
  address: string;
  name: string;
  symbol: string;
  totalSupply: number;
  floorPrice: string;
  holders: number;
  transfers: number;
  royaltyBps: number;
}

export interface Account {
  address: string;
  balance: string;
  transactionCount: number;
}

export interface NetworkStats {
  blockNumber: number;
  totalAddresses: number;
  totalTransactions: number;
  totalValueLocked: string;
  marketCap: string;
  tps: number;
  gasPrice: string;
}

export interface GasPrice {
  slow: string;
  standard: string;
  fast: string;
}

export interface ChartDataPoint {
  timestamp: number;
  value: number;
}

// SDK Configuration
export interface SDKConfig {
  apiKey?: string;
  baseURL?: string;
  chainId?: ChainId;
  timeout?: number;
}

// SDK Client
export class TigerScanSDK {
  private apiKey: string;
  private baseURL: string;
  private chainId: ChainId;
  private timeout: number;
  private http: AxiosInstance;

  constructor(config: SDKConfig = {}) {
    this.apiKey = config.apiKey || process.env.TIGERSCAN_API_KEY || '';
    this.baseURL = config.baseURL || 'https://api.tigerscan.io';
    this.chainId = config.chainId || ChainId.TIGER_MAINNET;
    this.timeout = config.timeout || 30000;

    this.http = axios.create({
      baseURL: this.baseURL,
      timeout: this.timeout,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'TigerScan-SDK-TypeScript/1.0.0',
        ...(this.apiKey && { 'X-API-Key': this.apiKey }),
      },
    });
  }

  // Block Operations

  async getBlock(blockNumber?: number, blockHash?: string): Promise<Block> {
    const endpoint = blockNumber !== undefined 
      ? `/api/v2/block/${blockNumber}`
      : `/api/v2/block/hash/${blockHash}`;
    
    const response = await this.http.get(endpoint);
    const result = response.data.result;
    
    return {
      number: parseInt(result.blockNumber),
      hash: result.hash,
      parentHash: result.parentHash,
      nonce: result.nonce,
      sha3Uncles: result.sha3Uncles,
      logsBloom: result.logsBloom,
      transactionsRoot: result.transactionsRoot,
      stateRoot: result.stateRoot,
      receiptsRoot: result.receiptsRoot,
      miner: result.miner,
      difficulty: parseInt(result.difficulty),
      totalDifficulty: parseInt(result.totalDifficulty),
      size: parseInt(result.size),
      gasLimit: parseInt(result.gasLimit),
      gasUsed: parseInt(result.gasUsed),
      timestamp: parseInt(result.timestamp),
      transactionCount: result.transactions?.length || 0,
      transactions: result.transactions || [],
      uncles: result.uncles || [],
    };
  }

  async getLatestBlock(): Promise<Block> {
    const response = await this.http.get('/api/v2/block/latest');
    return this.getBlock(parseInt(response.data.result.blockNumber));
  }

  async getBlockRewards(blockNumber: number): Promise<{ minerReward: string; uncleReward: string }> {
    const response = await this.http.get(`/api/v2/block/${blockNumber}/rewards`);
    return response.data.result;
  }

  // Transaction Operations

  async getTransaction(txHash: string): Promise<Transaction> {
    const response = await this.http.get(`/api/v2/tx/${txHash}`);
    const result = response.data.result;
    
    return {
      hash: result.hash,
      nonce: parseInt(result.nonce),
      blockHash: result.blockHash,
      blockNumber: parseInt(result.blockNumber),
      transactionIndex: parseInt(result.transactionIndex),
      from: result.from,
      to: result.to,
      value: result.value,
      gasPrice: result.gasPrice,
      gasLimit: parseInt(result.gasLimit),
      gasUsed: parseInt(result.gasUsed),
      input: result.input,
      status: parseInt(result.status),
      timestamp: parseInt(result.timestamp || 0),
    };
  }

  async getTransactionReceipt(txHash: string): Promise<TransactionReceipt> {
    const response = await this.http.get(`/api/v2/tx/${txHash}/receipt`);
    const result = response.data.result;
    
    return {
      transactionHash: result.transactionHash,
      blockHash: result.blockHash,
      blockNumber: parseInt(result.blockNumber),
      cumulativeGasUsed: parseInt(result.cumulativeGasUsed),
      gasUsed: parseInt(result.gasUsed),
      contractAddress: result.contractAddress,
      status: parseInt(result.status),
      logs: result.logs || [],
    };
  }

  async getInternalTransactions(txHash: string): Promise<any[]> {
    const response = await this.http.get(`/api/v2/tx/${txHash}/internal`);
    return response.data.result || [];
  }

  // Account Operations

  async getAccount(address: string): Promise<Account> {
    const response = await this.http.get(`/api/v2/account/${address}`);
    const result = response.data.result;
    
    return {
      address: result.address,
      balance: result.balance,
      transactionCount: parseInt(result.transactionCount),
    };
  }

  async getAccountTransactions(
    address: string, 
    page: number = 1, 
    limit: number = 50
  ): Promise<Transaction[]> {
    const response = await this.http.get(`/api/v2/account/${address}/transactions`, {
      params: { page, limit },
    });
    
    return (response.data.result || []).map((tx: any) => ({
      hash: tx.hash,
      nonce: parseInt(tx.nonce),
      blockHash: tx.blockHash,
      blockNumber: parseInt(tx.blockNumber),
      transactionIndex: parseInt(tx.transactionIndex),
      from: tx.from,
      to: tx.to,
      value: tx.value,
      gasPrice: tx.gasPrice,
      gasLimit: parseInt(tx.gasLimit),
      gasUsed: parseInt(tx.gasUsed),
      input: tx.input,
      status: parseInt(tx.status),
      timestamp: parseInt(tx.timestamp || 0),
    }));
  }

  async getAccountTokens(address: string): Promise<Token[]> {
    const response = await this.http.get(`/api/v2/account/${address}/tokens`);
    return (response.data.result || []).map((t: any) => ({
      address: t.address,
      name: t.name,
      symbol: t.symbol,
      decimals: parseInt(t.decimals),
      totalSupply: t.totalSupply,
      price: t.price,
      marketCap: t.marketCap,
      volume24h: t.volume24h,
      holders: parseInt(t.holders),
      transfers: parseInt(t.transfers),
    }));
  }

  async getAccountNFTs(address: string): Promise<NFT[]> {
    const response = await this.http.get(`/api/v2/account/${address}/nfts`);
    return (response.data.result || []).map((n: any) => ({
      address: n.address,
      name: n.name,
      symbol: n.symbol,
      totalSupply: parseInt(n.totalSupply),
      floorPrice: n.floorPrice,
      holders: parseInt(n.holders),
      transfers: parseInt(n.transfers),
      royaltyBps: parseInt(n.royaltyBps),
    })));
  }

  // Token Operations

  async getTokens(page: number = 1, limit: number = 50): Promise<Token[]> {
    const response = await this.http.get('/api/v2/tokens', {
      params: { page, offset: limit },
    });
    
    return (response.data.result || []).map((t: any) => ({
      address: t.address,
      name: t.name,
      symbol: t.symbol,
      decimals: parseInt(t.decimals),
      totalSupply: t.totalSupply,
      price: t.price,
      marketCap: t.marketCap,
      volume24h: t.volume24h,
      holders: parseInt(t.holders),
      transfers: parseInt(t.transfers),
    }));
  }

  async getToken(address: string): Promise<Token> {
    const response = await this.http.get(`/api/v2/token/${address}`);
    const t = response.data.result;
    
    return {
      address: t.address,
      name: t.name,
      symbol: t.symbol,
      decimals: parseInt(t.decimals),
      totalSupply: t.totalSupply,
      price: t.price,
      marketCap: t.marketCap,
      volume24h: t.volume24h,
      holders: parseInt(t.holders),
      transfers: parseInt(t.transfers),
    };
  }

  async getTokenHolders(
    address: string, 
    page: number = 1, 
    limit: number = 50
  ): Promise<TokenHolder[]> {
    const response = await this.http.get(`/api/v2/token/${address}/holders`, {
      params: { page, limit },
    });
    
    return (response.data.result || []).map((h: any) => ({
      address: h.address,
      balance: h.balance,
      percent: parseFloat(h.percent),
    }));
  }

  async getTokenTransfers(
    address: string, 
    page: number = 1, 
    limit: number = 50
  ): Promise<any[]> {
    const response = await this.http.get(`/api/v2/token/${address}/transfers`, {
      params: { page, limit },
    });
    return response.data.result || [];
  }

  async getTokenPriceHistory(address: string, days: number = 30): Promise<ChartDataPoint[]> {
    const response = await this.http.get(`/api/v2/token/${address}/history`, {
      params: { days },
    });
    
    return (response.data.result || []).map((p: any) => ({
      timestamp: p.timestamp,
      value: parseFloat(p.price),
    }));
  }

  async getHolderDistribution(address: string): Promise<any[]> {
    const response = await this.http.get(`/api/v2/token/${address}/holderdistribution`);
    return response.data.result || [];
  }

  // NFT Operations

  async getNFTs(page: number = 1, filter: string = 'erc721'): Promise<NFT[]> {
    const response = await this.http.get('/api/v2/nfts', {
      params: { page, filter },
    });
    
    return (response.data.result || []).map((n: any) => ({
      address: n.address,
      name: n.name,
      symbol: n.symbol,
      totalSupply: parseInt(n.totalSupply),
      floorPrice: n.floorPrice,
      holders: parseInt(n.holders),
      transfers: parseInt(n.transfers),
      royaltyBps: parseInt(n.royaltyBps),
    }));
  }

  async getNFT(address: string): Promise<NFT> {
    const response = await this.http.get(`/api/v2/nft/${address}`);
    const n = response.data.result;
    
    return {
      address: n.address,
      name: n.name,
      symbol: n.symbol,
      totalSupply: parseInt(n.totalSupply),
      floorPrice: n.floorPrice,
      holders: parseInt(n.holders),
      transfers: parseInt(n.transfers),
      royaltyBps: parseInt(n.royaltyBps),
    };
  }

  async getNFTHolders(address: string, page: number = 1): Promise<any[]> {
    const response = await this.http.get(`/api/v2/nft/${address}/holders`, {
      params: { page },
    });
    return response.data.result || [];
  }

  async getNFTTransfers(address: string, page: number = 1): Promise<any[]> {
    const response = await this.http.get(`/api/v2/nft/${address}/transfers`, {
      params: { page },
    });
    return response.data.result || [];
  }

  async getNFTFloorPrice(address: string): Promise<string> {
    const response = await this.http.get(`/api/v2/nft/${address}/floor`);
    return response.data.result?.floorPrice || '0';
  }

  async getNFTMetadata(address: string, tokenId: string): Promise<any> {
    const response = await this.http.get(`/api/v2/nft/${address}/metadata`, {
      params: { tokenId },
    });
    return response.data.result;
  }

  async getNFTTraits(address: string): Promise<any[]> {
    const response = await this.http.get(`/api/v2/nft/${address}/traits`);
    return response.data.result || [];
  }

  // Analytics

  async getNetworkStats(): Promise<NetworkStats> {
    const response = await this.http.get('/api/v2/stats');
    const s = response.data.result;
    
    return {
      blockNumber: parseInt(s.blockNumber),
      totalAddresses: parseInt(s.totalAddresses),
      totalTransactions: parseInt(s.totalTransactions),
      totalValueLocked: s.totalValueLocked,
      marketCap: s.marketCap,
      tps: parseFloat(s.tps),
      gasPrice: s.gasPrice,
    };
  }

  async getTPS(interval: string = '24h'): Promise<number> {
    const response = await this.http.get('/api/v2/stats/tps', {
      params: { interval },
    });
    return parseFloat(response.data.result?.tps || '0');
  }

  async getGasPrice(): Promise<GasPrice> {
    const response = await this.http.get('/api/v2/stats/gas');
    const g = response.data.result;
    
    return {
      slow: g.slow,
      standard: g.standard,
      fast: g.fast,
    };
  }

  async getTVL(): Promise<string> {
    const response = await this.http.get('/api/v2/stats/tvl');
    return response.data.result?.tvl || '0';
  }

  async getMarketCap(): Promise<string> {
    const response = await this.http.get('/api/v2/stats/marketcap');
    return response.data.result?.marketcap || '0';
  }

  async getRichList(page: number = 1): Promise<any[]> {
    const response = await this.http.get('/api/v2/stats/richlist', {
      params: { page },
    });
    return response.data.result || [];
  }

  async getTopTokens(page: number = 1): Promise<Token[]> {
    const response = await this.http.get('/api/v2/stats/toptokens', {
      params: { page },
    });
    
    return (response.data.result || []).map((t: any) => ({
      address: t.address,
      name: t.name,
      symbol: t.symbol,
      decimals: parseInt(t.decimals),
      totalSupply: t.totalSupply,
      price: t.price,
      marketCap: t.marketCap,
      volume24h: t.volume24h,
      holders: parseInt(t.holders),
      transfers: parseInt(t.transfers),
    }));
  }

  async getTopNFTs(page: number = 1): Promise<NFT[]> {
    const response = await this.http.get('/api/v2/stats/topnfts', {
      params: { page },
    });
    
    return (response.data.result || []).map((n: any) => ({
      address: n.address,
      name: n.name,
      symbol: n.symbol,
      totalSupply: parseInt(n.totalSupply),
      floorPrice: n.floorPrice,
      holders: parseInt(n.holders),
      transfers: parseInt(n.transfers),
      royaltyBps: parseInt(n.royaltyBps),
    }));
  }

  // Chart Data

  async getTPSChart(interval: string = '24h'): Promise<ChartDataPoint[]> {
    const response = await this.http.get('/charts/tps', {
      params: { interval },
    });
    
    const series = response.data.series?.[0];
    return (series?.data || []).map((d: any) => ({
      timestamp: new Date(d.timestamp).getTime(),
      value: d.value,
    }));
  }

  async getGasChart(interval: string = '24h'): Promise<any> {
    const response = await this.http.get('/charts/gas', {
      params: { interval },
    });
    return response.data;
  }

  async getTVLChart(interval: string = '30d'): Promise<ChartDataPoint[]> {
    const response = await this.http.get('/charts/tvl', {
      params: { interval },
    });
    
    const series = response.data.series?.[0];
    return (series?.data || []).map((d: any) => ({
      timestamp: new Date(d.timestamp).getTime(),
      value: d.value,
    }));
  }

  async getTokenPriceChart(
    address: string, 
    interval: string = '30d'
  ): Promise<ChartDataPoint[]> {
    const response = await this.http.get(`/charts/token/${address}`, {
      params: { interval },
    });
    
    const series = response.data.series?.[0];
    return (series?.data || []).map((d: any) => ({
      timestamp: new Date(d.timestamp).getTime(),
      value: d.value,
    }));
  }

  async getGasHeatmap(days: number = 7): Promise<any[]> {
    const response = await this.http.get('/charts/heatmap', {
      params: { days },
    });
    return response.data.heatmap || [];
  }

  // Search

  async search(query: string): Promise<any> {
    const response = await this.http.get('/api/v2/search', {
      params: { q: query },
    });
    return response.data.result;
  }

  // Contract Interaction

  async readContract(
    address: string, 
    method: string, 
    params: string[] = []
  ): Promise<string> {
    const response = await this.http.get(`/api/v2/contract/${address}/read`, {
      params: { method, params },
    });
    return response.data.result;
  }

  async writeContract(
    address: string,
    method: string,
    params: string[],
    fromAddress: string
  ): Promise<string> {
    const response = await this.http.post(`/api/v2/contract/${address}/write`, {
      method,
      params,
      from: fromAddress,
    });
    return response.data.result;
  }

  // Export

  async exportData(
    exportType: string,
    options: {
      address?: string;
      format?: string;
      startBlock?: number;
      endBlock?: number;
      limit?: number;
    } = {}
  ): Promise<any> {
    const response = await this.http.get('/export', {
      params: {
        type: exportType,
        ...options,
      },
    });
    
    if (options.format === 'csv') {
      return response.data;
    }
    return response.data.results || [];
  }
}

// Wallet Class
export class TigerScanWallet {
  private privateKey: string;
  private wallet: Wallet;
  private sdk: TigerScanSDK;

  constructor(privateKey: string, sdk?: TigerScanSDK) {
    this.privateKey = privateKey;
    this.wallet = new Wallet(privateKey);
    this.sdk = sdk || new TigerScanSDK();
  }

  get address(): string {
    return this.wallet.address;
  }

  async getNonce(): Promise<number> {
    const account = await this.sdk.getAccount(this.wallet.address);
    return account.transactionCount;
  }

  async sendTransaction(
    to: string,
    value: string = '0',
    gasLimit: number = 21000
  ): Promise<string> {
    const gasData = await this.sdk.getGasPrice();
    const gasPrice = BigNumber.from(gasData.standard);

    const tx = {
      to,
      value: BigNumber.from(value),
      gasLimit,
      gasPrice,
    };

    const signedTx = await this.wallet.signTransaction(tx);
    // In production, broadcast via RPC
    return signedTx.hash;
  }

  async transferToken(
    tokenAddress: string,
    to: string,
    amount: string
  ): Promise<string> {
    const token = new Contract(
      tokenAddress,
      ['function transfer(address to, uint256 amount)'],
      this.wallet
    );

    const tx = await token.transfer(to, BigNumber.from(amount));
    return tx.hash;
  }
}

// Utility Functions
export function createSDK(config?: SDKConfig): TigerScanSDK {
  return new TigerScanSDK(config);
}

export function createWallet(
  privateKey: string, 
  sdk?: TigerScanSDK
): TigerScanWallet {
  return new TigerScanWallet(privateKey, sdk);
}

// Export types
export type {
  Block,
  Transaction,
  TransactionReceipt,
  Log,
  Token,
  TokenHolder,
  NFT,
  Account,
  NetworkStats,
  GasPrice,
  ChartDataPoint,
};

// Export default
export default TigerScanSDK;

File: explorer/services/sdk/sdk.go
// Package sdk provides multi-language SDK support for the explorer
package sdk

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
)

// SDKClient represents an SDK client
type SDKClient struct {
	APIKey    string
	BaseURL   string
	ChainID  uint64
	httpClient *HTTPClient
	mu       sync.RWMutex
}

// HTTPClient represents HTTP client interface
type HTTPClient struct {
	baseURL string
	apiKey string
}

// NewSDKClient creates new SDK client
func NewSDKClient(apiKey, baseURL string, chainID uint64) *SDKClient {
	return &SDKClient{
		APIKey:   apiKey,
		BaseURL:  baseURL,
		ChainID: chainID,
		httpClient: &HTTPClient{
			baseURL: baseURL,
			apiKey:  apiKey,
		},
	}
}

// GetBlock gets block by number or hash
func (c *SDKClient) GetBlock(blockNumberOrHash string) (*Block, error) {
	data, err := c.httpClient.Get(fmt.Sprintf("/blocks/%s", blockNumberOrHash))
	if err != nil {
		return nil, err
	}
	
	var block Block
	if err := json.Unmarshal(data, &block); err != nil {
		return nil, err
	}
	
	return &block, nil
}

// GetTransaction gets transaction by hash
func (c *SDKClient) GetTransaction(txHash string) (*Transaction, error) {
	data, err := c.httpClient.Get(fmt.Sprintf("/transactions/%s", txHash))
	if err != nil {
		return nil, err
	}
	
	var tx Transaction
	if err := json.Unmarshal(data, &tx); err != nil {
		return nil, err
	}
	
	return &tx, nil
}

// GetAccount gets account balance
func (c *SDKClient) GetAccount(address string) (*Account, error) {
	data, err := c.httpClient.Get(fmt.Sprintf("/accounts/%s", address))
	if err != nil {
		return nil, err
	}
	
	var account Account
	if err := json.Unmarshal(data, &account); err != nil {
		return nil, err
	}
	
	return &account, nil
}

// GetToken gets token info
func (c *SDKClient) GetToken(address string) (*Token, error) {
	data, err := c.httpClient.Get(fmt.Sprintf("/tokens/%s", address))
	if err != nil {
		return nil, err
	}
	
	var token Token
	if err := json.Unmarshal(data, &token); err != nil {
		return nil, err
	}
	
	return &token, nil
}

// GetTokenTransfers gets token transfers
func (c *SDKClient) GetTokenTransfers(address string, opts *QueryOpts) ([]*Transfer, error) {
	endpoint := fmt.Sprintf("/tokens/%s/transfers", address)
	if opts != nil {
		endpoint = addQueryParams(endpoint, opts)
	}
	
	data, err := c.httpClient.Get(endpoint)
	if err != nil {
		return nil, err
	}
	
	var transfers []*Transfer
	if err := json.Unmarshal(data, &transfers); err != nil {
		return nil, err
	}
	
	return transfers, nil
}

// GetTransactions gets transactions
func (c *SDKClient) GetTransactions(opts *QueryOpts) ([]*Transaction, error) {
	endpoint := "/transactions"
	if opts != nil {
		endpoint = addQueryParams(endpoint, opts)
	}
	
	data, err := c.httpClient.Get(endpoint)
	if err != nil {
		return nil, err
	}
	
	var txs []*Transaction
	if err := json.Unmarshal(data, &txs); err != nil {
		return nil, err
	}
	
	return txs, nil
}

// GetInternalTransactions gets internal transactions
func (c *SDKClient) GetInternalTransactions(txHash string) ([]*InternalTx, error) {
	data, err := c.httpClient.Get(fmt.Sprintf("/transactions/%s/internal", txHash))
	if err != nil {
		return nil, err
	}
	
	var txs []*InternalTx
	if err := json.Unmarshal(data, &txs); err != nil {
		return nil, err
	}
	
	return txs, nil
}

// GetLogs gets logs
func (c *SDKClient) GetLogs(opts *LogQueryOpts) ([]*Log, error) {
	data, err := c.httpClient.Post("/logs", opts)
	if err != nil {
		return nil, err
	}
	
	var logs []*Log
	if err := json.Unmarshal(data, &logs); err != nil {
		return nil, err
	}
	
	return logs, nil
}

// GetContractABI gets contract ABI
func (c *SDKClient) GetContractABI(address string) (string, error) {
	data, err := c.httpClient.Get(fmt.Sprintf("/contracts/%s/abi", address))
	if err != nil {
		return "", err
	}
	
	return string(data), nil
}

// ReadContract reads from contract
func (c *SDKClient) ReadContract(address, method string, params []interface{}) (interface{}, error) {
	payload := map[string]interface{}{
		"to":   address,
		"data": method,
	}
	
	data, err := c.httpClient.Post("/contracts/read", payload)
	if err != nil {
		return nil, err
	}
	
	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}
	
	return result["result"], nil
}

// WriteContract writes to contract
func (c *SDKClient) WriteContract(address, method string, params []interface{}, options *TxOptions) (string, error) {
	payload := map[string]interface{}{
		"to":    address,
		"data":  method,
		"value": options.Value,
		"gas":   options.GasLimit,
	}
	
	data, err := c.httpClient.Post("/contracts/write", payload)
	if err != nil {
		return "", err
	}
	
	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return "", err
	}
	
	if hash, ok := result["hash"].(string); ok {
		return hash, nil
	}
	
	return "", fmt.Errorf("no transaction hash returned")
}

// EstimateGas estimates gas
func (c *SDKClient) EstimateGas(to, data string, value string) (uint64, error) {
	payload := map[string]interface{}{
		"to":   to,
		"data": data,
		"value": value,
	}
	
	dataBytes, err := c.httpClient.Post("/gas/estimate", payload)
	if err != nil {
		return 0, err
	}
	
	var result map[string]interface{}
	if err := json.Unmarshal(dataBytes, &result); err != nil {
		return 0, err
	}
	
	if gas, ok := result["gas"].(float64); ok {
		return uint64(gas), nil
	}
	
	return 21000, nil
}

// GetGasPrice gets current gas price
func (c *SDKClient) GetGasPrice() (*GasPrice, error) {
	data, err := c.httpClient.Get("/gas/prices")
	if err != nil {
		return nil, err
	}
	
	var gasPrice GasPrice
	if err := json.Unmarshal(data, &gasPrice); err != nil {
		return nil, err
	}
	
	return &gasPrice, nil
}

// GetBlockNumber gets current block number
func (c *SDKClient) GetBlockNumber() (uint64, error) {
	data, err := c.httpClient.Get("/blocks/latest/number")
	if err != nil {
		return 0, err
	}
	
	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return 0, err
	}
	
	if num, ok := result["number"].(float64); ok {
		return uint64(num), nil
	}
	
	return 0, fmt.Errorf("invalid block number")
}

// GetNFTs gets NFTs for address
func (c *SDKClient) GetNFTs(address string, opts *QueryOpts) ([]*NFT, error) {
	endpoint := fmt.Sprintf("/accounts/%s/nfts", address)
	if opts != nil {
		endpoint = addQueryParams(endpoint, opts)
	}
	
	data, err := c.httpClient.Get(endpoint)
	if err != nil {
		return nil, err
	}
	
	var nfts []*NFT
	if err := json.Unmarshal(data, &nfts); err != nil {
		return nil, err
	}
	
	return nfts, nil
}

// Search searches blocks, transactions, addresses
func (c *SDKClient) Search(query string) (*SearchResult, error) {
	data, err := c.httpClient.Get(fmt.Sprintf("/search?q=%s", query))
	if err != nil {
		return nil, err
	}
	
	var result SearchResult
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}
	
	return &result, nil
}

// GetTxReceipt gets transaction receipt
func (c *SDKClient) GetTxReceipt(txHash string) (*Receipt, error) {
	data, err := c.httpClient.Get(fmt.Sprintf("/transactions/%s/receipt", txHash))
	if err != nil {
		return nil, err
	}
	
	var receipt Receipt
	if err := json.Unmarshal(data, &receipt); err != nil {
		return nil, err
	}
	
	return &receipt, nil
}

// Subscribe subscribes to events (WebSocket)
func (c *SDKClient) Subscribe(ctx context.Context, channel string, handler func(interface{})) error {
	// WebSocket subscription would be implemented here
	return nil
}

// Types

type Block struct {
	Number       uint64        `json:"number"`
	Hash         string        `json:"hash"`
	ParentHash   string        `json:"parentHash"`
	Timestamp    uint64        `json:"timestamp"`
	Transactions []string      `json:"transactions"`
	GasUsed     uint64        `json:"gasUsed"`
	GasLimit    uint64        `json:"gasLimit"`
	Miner       string        `json:"miner"`
	Difficulty  string        `json:"difficulty"`
	TotalDifficulty string    `json:"totalDifficulty"`
	Size        uint64        `json:"size"`
}

type Transaction struct {
	Hash       string `json:"hash"`
	BlockNumber uint64 `json:"blockNumber"`
	BlockHash  string `json:"blockHash"`
	From      string `json:"from"`
	To        string `json:"to"`
	Value     string `json:"value"`
	GasPrice  string `json:"gasPrice"`
	GasLimit  uint64 `json:"gasLimit"`
	GasUsed   uint64 `json:"gasUsed"`
	Input     string `json:"input"`
	Nonce     uint64 `json:"nonce"`
	TxIndex  uint64 `json:"transactionIndex"`
	Timestamp uint64 `json:"timestamp"`
	Status   uint64 `json:"status"`
}

type Account struct {
	Address   string `json:"address"`
	Balance  string `json:"balance"`
	Nonce    uint64 `json:"nonce"`
	CodeHash string `json:"codeHash"`
	StorageRoot string `json:"storageRoot"`
}

type Token struct {
	Address       string `json:"address"`
	Name         string `json:"name"`
	Symbol       string `json:"symbol"`
	Decimals     uint8  `json:"decimals"`
	TotalSupply string `json:"totalSupply"`
	Type        string `json:"type"`
}

type Transfer struct {
	Hash       string `json:"hash"`
	From      string `json:"from"`
	To        string `json:"to"`
	Value     string `json:"value"`
	Token     string `json:"tokenAddress"`
	Timestamp uint64 `json:"timestamp"`
}

type InternalTx struct {
	Hash  string `json:"hash"`
	From  string `json:"from"`
	To    string `json:"to"`
	Value string `json:"value"`
	Type  string `json:"type"`
}

type Log struct {
	Address     string   `json:"address"`
	Topics     []string `json:"topics"`
	Data       string   `json:"data"`
	BlockNumber uint64  `json:"blockNumber"`
	TxHash    string   `json:"transactionHash"`
	LogIndex  uint64   `json:"logIndex"`
}

type NFT struct {
	TokenID     string `json:"tokenId"`
	Contract   string `json:"contractAddress"`
	Name       string `json:"name"`
	Symbol    string `json:"symbol"`
	URI       string `json:"tokenURI"`
	Owner     string `json:"owner"`
}

type GasPrice struct {
	Slow     string `json:"slow"`
	Standard string `json:"standard"`
	Fast    string `json:"fast"`
	BaseFee string `json:"baseFeePerGas"`
}

type Receipt struct {
	TransactionHash  string   `json:"transactionHash"`
	BlockNumber    uint64   `json:"blockNumber"`
	Status        uint64   `json:"status"`
	CumulativeGasUsed uint64 `json:"cumulativeGasUsed"`
	GasUsed       uint64   `json:"gasUsed"`
	Logs          []*Log   `json:"logs"`
}

type QueryOpts struct {
	Page     int `json:"page"`
	Limit   int `json:"limit"`
	Offset  int `json:"offset"`
	StartBlock uint64 `json:"startBlock"`
	EndBlock uint64 `json:"endBlock"`
}

type LogQueryOpts struct {
	Address   string   `json:"address"`
	Topics    []string `json:"topics"`
	FromBlock uint64   `json:"fromBlock"`
	ToBlock  uint64   `json:"toBlock"`
}

type TxOptions struct {
	Value    string `json:"value"`
	GasLimit uint64 `json:"gasLimit"`
	GasPrice string `json:"gasPrice"`
}

type SearchResult struct {
	Blocks   []*Block      `json:"blocks"`
	Transactions []*Transaction `json:"transactions"`
	Addresses   []string   `json:"addresses"`
	Tokens     []*Token    `json:"tokens"`
}

// Helper functions

func addQueryParams(endpoint string, opts *QueryOpts) string {
	var params []string
	if opts.Page > 0 {
		params = append(params, fmt.Sprintf("page=%d", opts.Page))
	}
	if opts.Limit > 0 {
		params = append(params, fmt.Sprintf("limit=%d", opts.Limit))
	}
	if opts.Offset > 0 {
		params = append(params, fmt.Sprintf("offset=%d", opts.Offset))
	}
	if opts.StartBlock > 0 {
		params = append(params, fmt.Sprintf("startblock=%d", opts.StartBlock))
	}
	if opts.EndBlock > 0 {
		params = append(params, fmt.Sprintf("endblock=%d", opts.EndBlock))
	}
	
	if len(params) > 0 {
		return endpoint + "?" + strings.Join(params, "&")
	}
	
	return endpoint
}

// Get is a placeholder for HTTP GET
func (h *HTTPClient) Get(endpoint string) ([]byte, error) {
	return []byte("{}"), nil
}

// Post is a placeholder for HTTP POST
func (h *HTTPClient) Post(endpoint string, data interface{}) ([]byte, error) {
	return []byte("{}"), nil
}

// InitSDK initializes the SDK
func InitSDK(apiKey, baseURL string, chainID uint64) (*SDKClient, error) {
	return NewSDKClient(apiKey, baseURL, chainID), nil
}